eng/fv/approaches.rst

```diff
@@ -0,0 +1,191 @@
+
+Given the emphasis on verifying RTEMS C code,
+we rapidly focussed in on freely available tools that could easily connect to C.
+These included: Frama-C, TLA+/PlusCal, Isabelle/HOL, and Promela/SPIN.
+Further investigation ruled out TLA+/PlusCal because it is Java-based,
+and requires installing a Java Runtime Environment.
+
+Frama-C, Isabelle/HOL, and Promela/SPIN were all explored in more detail,
+and used in initial experiments to assess their suitability.
+A key decision was also made early in the project to start looking at how to use
+formal techniques to do test generation.
+
+
+Frama-C
+^^^^^^^
+
+Frama-C (frama-c.com) is a platform supporting a range of tools for analysing C
+code, including static analysers, support for functional specifications (ANSI-C
+Specification Language – ACSL), and links to theorem provers. Some of its
+analyses require code annotations, while other can extract useful information
+from un-annotated code. It has a plug-in architecture, which makes it easy to
+extend. It is used extensively by Airbus.
+
+Frama-C, and its plugins, are implemented in OCaml,
+and it is installed using the ``opam`` package manager.
+An issue here was that Frama-C has many quite large dependencies.
+There was support for test generation, but it was not freely available.
+Another issue was that Frama-C only supported C99, and not C11
+(the issue is how to handle C11 Atomics in terms of their semantics).
+
+
+Isabelle/HOL
+^^^^^^^^^^^^
+
+Isabelle/HOL is a wide-spectrum theorem-prover, implemented as an embedding of
+Higher-Order Logic (HOL) into the Isabelle generic proof assistant
+(isabelle.in.tum.de). It has a high degree of automation, including an ability
+to link to third-party verification tools, and a very large library of verified
+mathematical theorems, covering number and set theory, algebra, analysis. It is
+based on the idea of a small trusted code kernel that defines an encapsulated
+datatype representing a theorem, which can only be constructed using methods in
+the kernel for that datatype, but which also scales effectively regardless of
+how many  theorems are proven.
+It is implemented using `polyml`, with the IDE implemented using Scala,
+is open-source, and is easy to install.
+However, like Frama-C, it is also a very large software suite.
+
+
+
+
+Formal Method actually used
+---------------------------
+
+A good survey of formal techniques and testing
+is found in a 2009 ACM survey paper :cite:`Hierons:2009:FMT`.
+Here they clearly state:
+
+  "The most important role for formal verification in testing
+  is in the automated generation of test cases.
+  In this context,
+  model checking is the formal verification technology of choice;
+  this is due to the ability of model checkers
+  to produce counterexamples
+  in case a temporal property does not hold for a system model."
+
+
+Promela/SPIN
+^^^^^^^^^^^^
+
+The current use of formal methods in RTEMS is based on using the Promela
```

Comment:
It may be useful to refer explicitly to "RTEMS pre-qualification" ? This is probably a general sub-project that needs to be defined, to encapsulate both any refernces to ESA-sponsored activity, and to all other activity related to qualification support provided via RTEMS Project.

Reply:
ok - will do